interp/src/stk_env.rs

```diff
@@ -0,0 +1,809 @@
+    /// share (references to the same) elements in their **tail**.
+    /// # Example
+    /// Good:
+    /// ```
+    /// use interp::stk_env::List;
+    /// let l1 = List::new().push(4).push(3);
+    /// if let (Some(hd), l1) = l1.split(){
+    ///     assert_eq!(hd, 3);
+    /// } else {
+    ///     panic!("could not split l1")
+    /// }
+    /// ```
+    /// Shared tail panic:
+    /// ```should_panic
+    /// use interp::stk_env::List;
+    /// let l1 = List::new().push(4).push(3);
+    /// let l2 = l1.push(2);
+    /// //This will panic, because while l1 exists, l2's tail cannot be split:
+    /// if let (Some(hd), l2) = l2.split() {
+    ///     let tup = l2.split();
+    /// }
+    /// l1.push(5);
+    /// ```
+    pub fn split(self) -> (Option<T>, Self) {
+        if self.is_empty() {
+            return (None, self);
+        } else {
+            if let Ok(head) = Rc::try_unwrap(self.head.unwrap()) {
+                let tail_list = List { head: head.next }; //better: not cloning the tail
+                return (Some(head.elem), tail_list);
+            } else {
+                panic!("Cannot unwrap the head of this list. You probably tried Smooshing while a fork exists!")
+            }
+        }
+    }
+
+    /// Returns an Option of a pointer to the head of this list
+    pub fn head(&self) -> Option<&T> {
+        self.head.as_ref().map(|node| &node.elem)
+    }
+
+    pub fn iter(&self) -> Iter<'_, T> {
+        Iter {
+            next: self.head.as_deref(),
+        }
+    }
+}
+
+pub struct Iter<'a, T> {
+    next: Option<&'a Node<T>>,
+}
+
+impl<'a, T> Iterator for Iter<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.next.map(|node| {
+            self.next = node.next.as_deref();
+            &node.elem
+        })
+    }
+}
+
+///A Stack of HashMaps that supports scoping
+///
+///Invariants:
+///=> To Smoosh X levels down, no fork may still exist from any node amongst those
+/// X levels. So the stack must look like:
+/// ```text
+///  _
+/// |_|
+///  |
+///  _
+/// |_|
+///  |
+///  _
+/// |_|
+///  |
+/// ```
+///=> To merge, both branches must share a common fork point
+/// ```text
+///  _
+/// |_|
+///  |
+///  _     _
+/// |_|   |_|
+///  |     |
+///   \ _ /
+///    |*|
+///     |
+/// ```
+#[derive(Debug)]
+pub struct Smoosher<K: Eq + std::hash::Hash, V: Eq> {
+    head: HashMap<K, V>,       //mutable
+    tail: List<HashMap<K, V>>, //read-only
+}
+
+impl<K: Eq + std::hash::Hash, V: Eq> Smoosher<K, V> {
+    pub fn new() -> Smoosher<K, V> {
+        Smoosher {
+            head: HashMap::new(),
+            tail: List::new(),
+        }
+    }
+
+    /// If [self] and [other] share a fork point, returns a pair (depthA, depthB)
+    /// of the depth which the fork point can be found in [self] and [other], respectively.
+    /// NOTE: should be private, only public for testing!
+    pub fn shared_fork_point(&self, other: &Self) -> Option<(u64, u64)> {
```

Comment:
You can write tests within this module (or make it pub(crate) so that it's not exposed beyond the boundaries of the interpreter).

See https://doc.rust-lang.org/book/ch11-03-test-organization.html#testing-private-functions